autotest/test_geospatial_index.py

```diff
@@ -0,0 +1,663 @@
+
+    result = grid.intersect(0.5, 0.5)
+    assert isinstance(result, (int, np.integer))
+
+    # Outside -> nan
+    result = index.query_point(10.0, 10.0)
+    assert np.isnan(result)
+
+    result = grid.intersect(10.0, 10.0, forgive=True)
+    assert np.isnan(result)
+
+
+@pytest.mark.parametrize(
+    "grid_fixture",
+    [
+        "minimal_vertex_grid",
+        "minimal_unstructured_grid",
+    ],
+)
+def test_return_type_array(grid_fixture, request):
+    """Test array return types: int64 for all inside, float64 when nan present."""
+    grid = request.getfixturevalue(grid_fixture)
+    index = GeospatialIndex(grid)
+
+    x_inside = np.array([0.5, 1.5])
+    y_inside = np.array([0.5, 0.5])
+    x_mixed = np.array([0.5, 10.0])
+    y_mixed = np.array([0.5, 10.0])
+
+    # All inside -> integer type
+    result = index.query_points(x_inside, y_inside)
+    assert np.issubdtype(result.dtype, np.integer)
+
+    result = grid.intersect(x_inside, y_inside)
+    assert np.issubdtype(result.dtype, np.integer)
+
+    # Mixed -> float64
+    result = index.query_points(x_mixed, y_mixed)
+    assert result.dtype == np.float64
+    assert not np.isnan(result[0])
+    assert np.isnan(result[1])
+
+    result = grid.intersect(x_mixed, y_mixed, forgive=True)
+    assert result.dtype == np.float64
+
+
+def test_return_type_structured(simple_structured_grid):
+    """Test StructuredGrid return types are consistent."""
+    grid = simple_structured_grid
+
+    # Scalar inside -> int
+    row, col = grid.intersect(5.0, 95.0)
+    assert isinstance(row, (int, np.integer))
+    assert isinstance(col, (int, np.integer))
+
+    # Scalar outside -> nan
+    row, col = grid.intersect(150.0, 50.0, forgive=True)
+    assert np.isnan(row) and np.isnan(col)
+
+    # Scalar with z inside -> int
+    lay, row, col = grid.intersect(5.0, 95.0, z=5.0)
+    assert all(isinstance(v, (int, np.integer)) for v in [lay, row, col])
+
+    # Scalar with z outside -> nan
+    lay, row, col = grid.intersect(150.0, 50.0, z=5.0, forgive=True)
+    assert all(np.isnan(v) for v in [lay, row, col])
+
+    # Array all inside -> any integer dtype
+    rows, cols = grid.intersect(np.array([5.0, 55.0]), np.array([95.0, 95.0]))
+    assert np.issubdtype(rows.dtype, np.integer)
+    assert np.issubdtype(cols.dtype, np.integer)
+
+    # Array with outside -> must be float64 (due to NaNs)
+    rows, cols = grid.intersect(
+        np.array([5.0, 150.0]), np.array([95.0, 50.0]), forgive=True
+    )
+    assert rows.dtype == np.float64
+    assert cols.dtype == np.float64
+
+
+# ============================================================================
+# Test Edge Cases
+# ============================================================================
+
+
+def test_thin_sliver_cell():
+    """Test GeospatialIndex finds points in very thin "sliver" cells.
+
+    Tests the centroid+vertices KD-tree approach for cells where the
+    centroid might be far from the actual cell location.
```

Comment:
what do you mean by "centroid might be far from the actual cell location"? the centroid of a triangular "sliver" cell will still be in the cell, right? wouldn't the cell have to be e.g. a crescent for the centroid to fall outside of it?